overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+	}
+
+	/// Send a direct message to some other `Subsystem` you know `SubsystemId` of.
+	pub async fn send_msg_to(&mut self, to: SubsystemId, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
+			to: Some(to),
+			msg,
+		}).await;
+	}
+
+	fn new(rx: mpsc::Receiver<M>, tx: mpsc::Sender<OverseerMessage<M>>) -> Self {
+		Self {
+			rx,
+			tx,
+		}
+	}
+}
+
+/// A trait that describes the `Subsystems` that can run on the `Overseer`.
+///
+/// It is generic over the message type circulating in the system.
+/// The idea that we want some type contaning persistent state that
+/// can start actually running jobs when asked to.
+pub trait Subsystem<M: Debug> {
+	/// Start this `Subsystem` and return `SubsystemJob`.
+	fn start(&mut self, ctx: SubsystemContext<M>) -> SubsystemJob;
+	/// If this `Subsystem` want to receive this message.
+	///
+	/// By default receive all messages.
+	fn can_recv_msg(&self, _msg: &M) -> bool { true }
+}
+
+/// A subsystem that we oversee.
+///
+/// Ties together the `Subsystem` itself and it's running instance
+/// (which may be missing if the `Subsystem` is not running at the moment
+/// for whatever reason).
+struct OverseedSubsystem<M: Debug> {
+	subsystem: Box<dyn Subsystem<M>>,
+	instance: Option<SubsystemInstance<M>>,
+}
+
+/// The `Overseer` itself.
+pub struct Overseer<M: Debug> {
+	/// All `Subsystem`s by their respective `SubsystemId`s.
+	subsystems: HashMap<SubsystemId, OverseedSubsystem<M>>,
+	/// The actual poor man's process tree.
+	///
+	/// Needed (among other things) to stop a running `Job` along
+	/// with all it's children.
+	id_to_children: HashMap<SubsystemId, HashSet<SubsystemId>>,
+}
+
+impl<M: Debug + Clone> Overseer<M> {
+	/// Create a new intance of the `Overseer` with some initial set of `Subsystems.
+	///
+	/// The `Subsystems` submitted to this call will act as a level 1 in the "process tree":
+	///
+	///
+	/// ```text
+	///                  +------------------------------------+
+	///                  |            Overseer                |
+	///                  +------------------------------------+
+	///                    /            |             |      \
+	///      ................. subsystems[..] ..............................
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      . |           |    |           |   |          |   |         | .
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      ...............................................................
+	///                              |
+	///                        probably `spawn`
+	///                        something else
+	///                              |
+	///                              V
+	///                         +-----------+
+	///                         |           |
+	///                         +-----------+
+	///
+	/// ```
+	pub fn new(subsystems: Vec<Box<dyn Subsystem<M>>>) -> Self {
```

Comment:
I would rather have a load of generic arguments than type-erasure, as it makes things more complicated.